### ISSUE TITLE
pip install fails because of sphinx 4.0.0b1

### ISSUE THREAD

**Issue by kolewu** [BobBuildTool/bob#408]
When the prebuilt wheel is missing (or not allowed) the build of 0.18.0, 0.18.1 and 0.19.0 fails if sphinx setup dependency is resolved to v4.0 prerelease.

This could be easily tested with this `Dockerfile`:

```Dockerfile
FROM debian:buster

RUN set -ex; \
	apt-get update && \
	apt-get -y install python3-pip && \
	rm -rf /var/lib/appt/lists/*

# fails without 'python3-sphinx' installed
RUN set -ex; \
	python3 -m pip install --no-binary BobBuildTool BobBuildTool==0.19.0
```

I think the quick solution is to forbid 4.0 for now in `setup.py` (but I'm not sure about the syntax):
```diff
     build.sub_commands.append(('build_sphinx', None))
     setup_requires.extend([
-        'sphinx',
+        'sphinx<4.0',
     ])

**Comment by jkloetzke** [comment]
I'll create a 0.18.2 and 0.19.1 release that pins the dependencies to their major versions. Hopefully this will prevent such install failure in the future...

**Comment by kolewu** [comment]
If you pin to the major version (3 in this case), most of the time the installed sphinx version cannot be used. Debian Buster and openSUSE Leap 15.3 are still at 1.8. Since only the new version does not work, I suggested to accept all versions prior to 4.0.

**Comment by kolewu** [comment]
Just forgot to mention the workaround: Install sphinx (with a version smaller than 4.0) before installing bob makes the buildscript happy.

**Comment by jkloetzke** [comment]
Yes, you're right.  OTOH I was just trying to reproduce locally and here the pre-release version is not used by default. I could only see a problem when explicitly installing the pre-relase version. I think some older pip versions incorrectly installed pre-release versions which might be why you already see the problem now.

Anyway, we should probably restrict the major versions of all dependencies to the known good versions that were usable at the time of a release. The master branch should work with the latest versions and would have no restrictions, though.

**Comment by jkloetzke** [comment]
This Python dependency handling is really f****d up. It pretty much guarantees that older packages will never be installable again. If I pin the dependencies of released versions you cannot install it globally because, depending on the distribution, there will already be conflicts. Limiting the dependencies to the latest tested major version looks better but most packages give a sh*t about semantic versioning. Even if it would work for Bobs direct dependencies, the transitive ones will still be floating and are guaranteed to break in the future. I'm not sure if there is any reasonable way how to solve this...

**Comment by jkloetzke** [comment]
I've made the master branch compatible to Sphinx 4. I'll queue the fix for 0.19.1. Or would you also need a new release of the 0.18 train too?

**Comment by kolewu** [comment]
Thanks, but no, the workaround is sufficient for the old releases.
